spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestMetadataTableReadableMetrics.java

```diff
@@ -0,0 +1,498 @@
+    expectedNanValues.put("decimalCol", null);
+    expectedNanValues.put("binaryCol", null);
+    expectedNanValues.put("fixedCol", null);
+    expectedNanValues.put("dateCol", null);
+    expectedNanValues.put("timeCol", null);
+    expectedNanValues.put("timestampCol", null);
+
+    checkMetricValues(row, "nan_value_count", expectedNanValues);
+  }
+
+  @Test
+  public void testNestedValues() throws Exception {
+    String tableName = "testNestedValues";
+    Table table = createTable(tableName, NESTED_SCHEMA, PartitionSpec.unpartitioned());
+
+    List<Record> records =
+        Lists.newArrayList(
+            createNestedRecord(0L, 0.0),
+            createNestedRecord(1L, Double.NaN),
+            createNestedRecord(null, null));
+    DataFile dataFile =
+        FileHelpers.writeDataFile(table, Files.localOutput(temp.newFile()), records);
+    table.newAppend().appendFile(dataFile).commit();
+
+    Dataset<Row> df = spark.read().format("iceberg").load("default." + tableName + ".files");
+    List<Row> rows = df.collectAsList();
+    Assert.assertEquals("Expected only one data file", 1, rows.size());
+    Row row = rows.get(0);
+
+    String[] nestedColumns =
+        new String[] {
+          "nestedStructCol.leafStructCol.leafDoubleCol", "nestedStructCol.leafStructCol.leafLongCol"
+        };
+    checkMetric(
+        row,
+        "column_size",
+        nestedColumns,
+        l -> Assert.assertTrue("Column size should be greater than 0", l > 0));
+    checkMetric(
+        row,
+        "value_count",
+        nestedColumns,
+        l -> Assert.assertEquals("Value count should be 3", l.longValue(), 3L));
+    checkMetric(
+        row,
+        "null_value_count",
+        nestedColumns,
+        l -> Assert.assertEquals("Null value count should be 1", l.longValue(), 1L));
+
+    Map<String, Long> expectedNanValues = Maps.newHashMap();
+    expectedNanValues.put("nestedStructCol.leafStructCol.leafDoubleCol", 1L);
+    expectedNanValues.put("nestedStructCol.leafStructCol.leafLongCol", null);
+    checkMetricValues(row, "nan_value_count", expectedNanValues);
+
+    checkMetricValues(
+        row,
+        "lower_bound",
+        ImmutableMap.of(
+            "nestedStructCol.leafStructCol.leafLongCol", "0",
+            "nestedStructCol.leafStructCol.leafDoubleCol", "0.0"));
+    checkMetricValues(
+        row,
+        "upper_bound",
+        ImmutableMap.of(
+            "nestedStructCol.leafStructCol.leafLongCol", "1",
+            "nestedStructCol.leafStructCol.leafDoubleCol", "0.0"));
+  }
+
+  private <T> void checkCollectionValues(Row row, String columnName, Set<T> expectedValues) {
+    Set<Long> actualValues =
+        Sets.newHashSet(JavaConverters.asJavaCollection(row.getAs(columnName)));
+    Assert.assertEquals("Collection values should match", expectedValues, actualValues);
+  }
+
+  private void checkMetric(Row row, String metricName, String[] columns, Consumer<Long> check) {
```

Comment:
So this is for checking if a certain metric is present for all columns and then run an arbitrary long function on that value?

I think it may help to break this into a function that reports missing metrics and a function which checks if a boolean expression is true for all such values. Currently we end up passing in check functions that have to implement their own assertions which seems a little confusing to me. 

Another option maybe something like

```
  private <T> void checkMetricValueForAllColumns(Row row, String metricName, String[] columns, Predicate<T> check) {
    scala.collection.Map<String, Row> metrics = row.getAs("readable_metrics");
    for (String column : columns) {
      Option<Row> rowOption = metrics.get(column);
      Assert.assertTrue("Missing metric for column: " + column, rowOption.isDefined());
      T metric = rowOption.get().getAs(metricName);
      Assert.assertTrue("Check on metric ....", check.test(metric));
    }
  }
```

Reply:
Got rid of these custom methods, used assertEquals() and sql() for everything